### ISSUE TITLE
Storing 0.9999999999999999 in DBDecimal results in storing 0

### ISSUE THREAD

**Issue by sb-relaxt-at** [silverstripe/silverstripe-framework#10592]
## Affected Version

Silverstripe 4.11 (Framework 4.11.10)

## Description

Saving a value of `0.9999999999999999` into a Decimal field will result in storing `0` into the field.

## Reason
In DBDecimal#prepValueForDB there is a rather strange logic:

````php
        if (ctype_digit((string) $value)) {
            return (int)$value;
        }
```` 

Given that `$value = 0.9999999999999999` the string representation will be `1`, therefore the if statement will evaluate to true. In turn the value will be casted to int, resulting in `0` being returned and stored. (Fun fact, depending on the PHP version a dump of the float value will be either `float(0.9999999999999999)` or `float(1)`, but the string representation is always `1`).

## Solution

I would suggest a check like this (requires PHP 7.2+): 

````php
        if (abs((float)$value - (int)$value) < PHP_FLOAT_EPSILON) {
            return (int)$value;
        }
```` 

_Edit: Of course, this is not limited to this specific value, e.g.:_

````php
$x = 7.999999999999999;
var_dump((string)$x, (int)$x); // "8", 7
````

## Acceptance criteria
- [ ] Numbers with very long decimal are rounded out properly when saved in an DB_Decimal field.

## PRs
- #10593

**Comment by michalkleiner** [comment]
Happy to look at a FIX PR against the 4.11 branch if you open one.

**Comment by michalkleiner** [comment]
Ok, started a [PR](https://github.com/silverstripe/silverstripe-framework/pull/10593) myself, first adding a test to see if it fails then will try with the suggested fix.

**Comment by emteknetnz** [comment]
Linked PR has been merged and released as [4.11.15](https://github.com/silverstripe/silverstripe-framework/releases/tag/4.11.15)

**Comment by sb-relaxt-at** [comment]
@michalkleiner Thanks a lot for the fast fix!
